README.md
[![CocoaPods](https://img.shields.io/cocoapods/v/DYAlertController.svg?style=flat)](http://cocoadocs.org/docsets/DYAlertController)
[![license](https://img.shields.io/github/license/DominikButz/DYAlertController.svg?style=flat)](http://cocoadocs.org/docsets/DYAlertController)
[![CocoaPods](https://img.shields.io/cocoapods/dt/DYAlertController.svg)]()



# DYAlertController

DYAlertController can be used as a replacement for Apple’s UIAlertController. 
It supports checkmarks, single and multiple selection and icons. DYAlertController also features two styles, alert and actionSheet. Tapping an action and tapping the ok or cancel button will trigger actions you define in the action item’s handler or in the cancel and ok button handlers, similar to UIAlertController.
Pull the framework and checkout the example project in the subfolder. 

Updated to the latest **Swift 3** syntax!


## Installation


Install DYAlertController through Cocoapods. Enter the following information into your Podfile (see current version in header):

```Ruby
platform :ios, '10.0'

use_frameworks!

target '[Your app project title]' do
pod 'DYAlertController', '~> [current version - see header]'
end

```
Make sure to import DYAlertController into your View Controller subclass:

```Swift
	import DYAlertController
```
**Important**: The Xcode compiler might show an error after you open your project .xcworkspace file including DYAlertController for the first time (something like "No such module DYAlertController"). I have had this issue with a lot of other Cocoapods which I tried to install before. Simply run your code and the error should disappear. 

Alternatively, you can pull this framework and copy the DYAlertController folder (in the pods folder) from the example project into your project.

 

##Features

As alternative to UIAlertController, DYAlertController has the following additional features:

* Add an icon image to the title view right above the title
* Add an icon image to an action
* If you add an ok button (which is optional), clicking on an action will not dismiss the alert or action sheet but will toggle a checkmark instead. You can also set the controller to multiple selection. If you don’t add an ok button action, the alert or action sheet will be dismissed when tapping an action
* change the ok button style (normal, destructive, disabled) in your action item handlers
* Add several text fields
* Choose from two background effect view styles, blur and dim
* Set a custom width for the alert or action sheet (its height will be set automatically depending on the content view’s subviews)
* Customise colours, fonts, corner radius etc.



## Usage

The usage is similar to UIAlertController. See the following example.


###Code example: Creating an alert

```Swift
let titleImage = UIImage(named: "shareIcon")
let alert = DYAlertController(style: .alert, title: "Doing stuff", titleIconImage: titleImage, message:"Select one option", cancelButtonTitle: "Cancel", multipleSelection: false, customFrameWidth:200, backgroundEffect: DYAlertController.EffectViewMode.blur)

    
alert.addAction(DYAlertAction(title: "Do stuff 1", style:.normal, iconImage: nil, setSelected:false, handler: { (alertAction) -> Void in
    
  print("executing first action! selected: \(alertAction.selected)")
                
}))
    
alert.addAction(DYAlertAction(title: "Do stuff 2", style:.normal, iconImage: nil, setSelected:false, handler: { (alertAction) -> Void in
    
  print("executing 2nd action! selected: \(alertAction.selected)")
    
}))
    
    
alert.addAction(DYAlertAction(title: "Beware!", style:.destructive, iconImage: nil, setSelected:true, handler: { (alertAction) -> Void in
 
    print("executing 3rd action! selected: \(alertAction.selected)")
}))

alert.handleCancelAction = {
    
    print("cancel tapped")
}
    
self.present(alert, animated: true, completion: nil)
    


```

![Alert example 1](https://github.com/DominikButz/DYAlertController/blob/master/gitResources/AlertExample1.gif "Alert example 1")





### Adding text fields
```Swift
 	let textfield1 = UITextField()
	textfield1.placeholder = "username"
	textfield1.autocorrectionType = UITextAutocorrectionType.no
	  
	let textfield2 = UITextField()
	textfield2.placeholder = "password"
	textfield2.isSecureTextEntry = true
	  
	let textfield3 = UITextField()
	textfield3.placeholder = "hostname"
	textfield3.autocorrectionType = UITextAutocorrectionType.no
	  
	textfield1.delegate = self
	textfield2.delegate = self
	textfield3.delegate = self
	  
	alert.addTextField(textField: textfield1)
	alert.addTextField(textField: textfield2)
	alert.addTextField(textField: textfield3)

```

You can get the text input like so: alert.textFields[0].text etc. 
Make sure to only add text fields to an alert, not to an action sheet - just like UIAlertController, **your app will crash at runtime if you try to add a text field to an action sheet**.

![Alert example 2](https://github.com/DominikButz/DYAlertController/blob/master/gitResources/AlertExample2.gif "Alert example 2")

### Adding an ok button action

Add an ok button action as follows:

```Swift
alert.addOKButtonAction("OK", setDisabled: false) { 
   print("ok button tapped!")
}
```
You can set the ok button to disabled state initially (e.g. if the user should change the selection first before he can tap the ok button. 
The button style can be changed in the action handlers. For example:

```Swift

 let actionSheet = DYAlertController(style: .actionSheet, title: "Doing stuff", titleIconImage: titleImage, message:"Select one option", cancelButtonTitle: "Cancel", multipleSelection: false, customFrameWidth:nil, backgroundEffect:.dim)
    
            
actionSheet.addAction(DYAlertAction(title: "Option 1", style:.normal, iconImage: UIImage(named: "eyeIcon"), setSelected:true, handler: { (action) -> Void in
    
    if action.selected {

      actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)
      

     } else {
       // this action is deselected
                    
       // check if all actions are deselected
       if actionSheet.allActionsDeselected() {
                        
           // let's disable the OK button if all actions are deselected
          actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
           }
                    
     }
                
  print("changing state of first option.  selected: \(action.selected)")
            
}))


// ... add other actions...


actionSheet.addOKButtonAction("OK", setDisabled: false) {
                
    var selectedOptionIndex:Int?
                
                // we had set the multipleSelection option to false, so only one can be selected
     for i in 0...actionSheet.alertActions.count-1 {
         if actionSheet.alertActions[i].selected {
             selectedOptionIndex = i
            }
      }
            
      if let _ = selectedOptionIndex {
                     print("ok button tapped -  option selected: \(actionSheet.alertActions[selectedOptionIndex!].title)")
       } else {
                    print("no option was selected!")  
// to play safe... but actually not possible because ok button set to disabled if none selected
        }
                
 }
            
 // ...

```
Download the example project for more details and check out the examples as shown in the gif animations below. 

### Action sheet examples

Action sheet with simple selection:

![Action Sheet example 1](https://github.com/DominikButz/DYAlertController/blob/master/gitResources/ActionSheetExample1.gif "ActionSheet example 1")


Action sheet with multiple selection:

![Action Sheet example 2](https://github.com/DominikButz/DYAlertController/blob/master/gitResources/ActionSheetExample2.gif "ActionSheet example 2")

### Customising
If you intend to create several alerts or action sheets with the same fonts and colours in your app you can simply change the default settings in the DYAlertSettings struct. Alternatively, you can change single properties in code by changing the properties in the struct instances. For example:

```Swift
// create alert or action sheet first
// ... 

alert.settings.contentViewCornerRadius = 2.0

alert.settings.titleTextColor = UIColor.black

```

These changes only overwrite the settings properties of your current DYAlertViewController instance. Check out the DYAlertSettings.swift file. 

## License

The MIT License (MIT)

Copyright (c) 2016 Dominik Butz

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.



CHANGELOG.md
# Change log

## [Version 1.0.6](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.6)
Released on 2017-01-03.
- action sheet animation replaced by spring with damping effect 

## [Version 1.0.5](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.5)
Released on 2017-01-03.
- updated testing 

## [Version 1.0.4](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.4)
Released on 2016-11-23.
- added testing (experimental)

## [Version 1.0.3](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.3)
Released on 2016-11-22.
- adding quick help documentation (DYAlertSettings)

## [Version 1.0.2](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.2)
Released on 2016-11-21.
- adding quick help documentation


## [Version 1.0.1](https://github.com/DominikButz/DYAlertController/releases/tag/1.0.1)
Released on 2016-11-18.
- fixing 'jump up' bug when orientation changes during presentation.


## [Version 1.0](https://github.com/DominikButz/DYAlertController/releases/tag/1.0)
Released on 2016-11-17.

- add several text fields

##[Version 0.6](https://github.com/DominikButz/DYAlertController/releases/tag/0.6)
Released on 2016-11-16

**Swift 3 support**



DYAlertControllerExample3/DYAlertControllerExample3/ViewController.swift
//
//  ViewController.swift
//  DYAlertControllerExample3
//
//  Created by Dominik Butz on 16/03/16.
//  Copyright © 2016 Duoyun. All rights reserved.
//

import UIKit
import DYAlertController


class ViewController: UIViewController, UITextFieldDelegate {

    override func viewDidLoad() {
        super.viewDidLoad()
        

    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
        @IBAction func alertExample1Tapped(_ sender: UIButton) {
    
    
            let titleImage = UIImage(named: "shareIcon")
            let alert = DYAlertController(style: .alert, title: "Doing stuff", titleIconImage: titleImage, message:"Select one option", cancelButtonTitle: "Cancel", multipleSelection: false, customFrameWidth:nil, backgroundEffect: DYAlertController.EffectViewMode.blur)

        
            alert.addAction(DYAlertAction(title: "Do stuff 1", style:.normal, iconImage: nil, setSelected:false, handler: { (alertAction) -> Void in
    
               print("executing first action! selected: \(alertAction.selected)")
                
            }))
    
            alert.addAction(DYAlertAction(title: "Do stuff 2", style:.normal, iconImage: nil, setSelected:false, handler: { (alertAction) -> Void in
    
               print("executing 2nd action! selected: \(alertAction.selected)")
    
            }))
    
    
            alert.addAction(DYAlertAction(title: "Beware!", style:.destructive, iconImage: nil, setSelected:false, handler: { (alertAction) -> Void in
    
    
               print("executing 3rd action! selected: \(alertAction.selected)")
    
            }))

            alert.handleCancelAction = {
    
               print("cancel tapped")
            }
    

            self.present(alert, animated: true, completion: nil)
    
        }
    
    
    
        @IBAction func alertExample2Tapped(_ sender: UIButton) {
    
            let titleImage = UIImage(named: "shareIcon")
            let alert = DYAlertController(style: .alert, title: "Login", titleIconImage: titleImage, message: "Enter your login data", cancelButtonTitle: "Cancel", multipleSelection: false, customFrameWidth:180.0, backgroundEffect: DYAlertController.EffectViewMode.dim)
    
            
            
            let textfield1 = UITextField()
            textfield1.placeholder = "username"
            textfield1.autocorrectionType = UITextAutocorrectionType.no
            
            let textfield2 = UITextField()
            textfield2.placeholder = "password"
            textfield2.isSecureTextEntry = true
            
            let textfield3 = UITextField()
            textfield3.placeholder = "hostname"
            textfield3.autocorrectionType = UITextAutocorrectionType.no
            
            textfield1.delegate = self
            textfield2.delegate = self
            textfield3.delegate = self
            
            alert.addTextField(textField: textfield1)
            alert.addTextField(textField: textfield2)
            alert.addTextField(textField: textfield3)
    
            alert.handleCancelAction = {
    
                print("cancel tapped")
            }
    
            alert.addOKButtonAction("OK", setDisabled: false) {
                
                
                 print("ok button with title \(alert.okButton?.titleLabel?.text) tapped")
                print("uername: \(alert.textFields[0].text!)")
                print("password: \(alert.textFields[1].text!)")
                print("hostname:\(alert.textFields[2].text!)")
            }
            
            
    
            self.present(alert, animated: true, completion: nil)
    
        }
    
    
        @IBAction func actionSheetExample1Tapped(_ sender: UIButton) {
    
            let titleImage = UIImage(named: "shareIcon")
            let actionSheet = DYAlertController(style: .actionSheet, title: "Doing stuff", titleIconImage: titleImage, message:"Select one option", cancelButtonTitle: "Cancel", multipleSelection: false, customFrameWidth:nil, backgroundEffect:.dim)
    
            
            actionSheet.addAction(DYAlertAction(title: "Option 1", style:.normal, iconImage: UIImage(named: "eyeIcon"), setSelected:true, handler: { (action) -> Void in
    
                if action.selected {

                    actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)
      

                } else {
                    // deselected
                    
                    if actionSheet.allActionsDeselected() {
                 
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                    }
                    
                    
                    
                }
                
                print("changing stat of first option.  selected: \(action.selected)")
            }))
    
            actionSheet.addAction(DYAlertAction(title: "Option 2", style:.normal, iconImage: UIImage(named: "locationIcon"), setSelected:false, handler: { (action) -> Void in
    
                if action.selected {
                
                    actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)
                } else {
                    // deselected
                    if actionSheet.allActionsDeselected() {
           
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                    }
                    
                }
                
               print("changing state of 2nd option. selected: \(action.selected)")
                
                
    
            }))
    
    
            actionSheet.addAction(DYAlertAction(title: "Option 3 - risky", style: .destructive, iconImage: UIImage(named: "eyeIcon"), setSelected:false, handler: { (action) -> Void in
                
                if action.selected {
          
                    actionSheet.okButton!.setDestructiveStyle("Beware!", titleColor: actionSheet.settings.okButtonTintColorDestructive)
                } else {
                    // deselected
                    if actionSheet.allActionsDeselected() {
            
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                    }
                    
                }
    
                        print("changing state of 3rd option. selected: \(action.selected)")
    
            }))
    

            actionSheet.addOKButtonAction("OK", setDisabled: false) {
                
                var selectedOptionIndex:Int?
                
                // we set the multipleSelection option to false, so only one can be selected
                for i in 0...actionSheet.alertActions.count-1 {
                    if actionSheet.alertActions[i].selected {
                        selectedOptionIndex = i
                    }
                }
            
                if let _ = selectedOptionIndex {
                     print("ok button tapped -  option selected: \(actionSheet.alertActions[selectedOptionIndex!].title)")
                } else {
                    print("no option was selected!")  // to play safe... but actually not possible because ok button set to disabled if none selected
                }
                
            }
            
            actionSheet.handleCancelAction = {
                
                print("cancel tapped")
            }
    
          
    
            self.present(actionSheet, animated: true, completion: nil)
        }
    
    
    
        @IBAction func actionSheetExample2Tapped(_ sender: UIButton) {
    
    
            let actionSheet = DYAlertController(style: .actionSheet, title: "Doing stuff", titleIconImage: nil, message:"Select one or several options", cancelButtonTitle: "Cancel", multipleSelection: true, customFrameWidth:200.0, backgroundEffect: .blur)
    
    
            actionSheet.addAction(DYAlertAction(title: "Option 1", style:.normal, iconImage: UIImage(named: "editIcon"), setSelected:false, handler: { (action) -> Void in
                
                if action.selected {
                    actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)

                } else {
                
                    if actionSheet.allActionsDeselected() {
          
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                    }
                }


            }))
    
            actionSheet.addAction(DYAlertAction(title: "Option 2", style:.normal, iconImage: UIImage(named: "locationIcon"), setSelected:false, handler: { (action) -> Void in
    
                print("changing state of 2nd option. selected: \(action.selected), activating OK button")
                if action.selected {
                  actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)
             
                } else {
                   if actionSheet.allActionsDeselected() {
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                    }
                }
                
            }))
    
    
            actionSheet.addAction(DYAlertAction(title: "Risky", style:.destructive, iconImage: UIImage(named: "eyeIcon"), setSelected:false, handler: { (action) -> Void in
                
                if action.selected {
                    actionSheet.okButton!.setDestructiveStyle("Beware!", titleColor:actionSheet.settings.okButtonTintColorDestructive)
                    
                } else {
                     if actionSheet.allActionsDeselected() {
                        actionSheet.okButton!.setDisabledStyle("Disabled", titleColor: actionSheet.settings.okButtonTintColorDisabled)
                     } else {
                        actionSheet.okButton!.setNormalStyle("OK", titleColor: actionSheet.settings.okButtonTintColorDefault)
                    }
                }
             
                
            }))
            
            actionSheet.handleCancelAction = {
                
               print("cancel tapped")
            }
            
            actionSheet.addOKButtonAction("OK", setDisabled: true) {
                      // can't be tapped when sheet starts up because of state 'disabled' but can be  changed through user selection - see actions above
            }

            
            
            self.present(actionSheet, animated: true, completion: nil)
            
        }
    

    func textFieldDidBeginEditing(_ textField: UITextField) {
        print("began editing!")
        
        if textField.placeholder == "username" {
            print("editing username field!")
        }
    }


}

